src/backend/dx12/src/descriptors_cpu.rs

```diff
@@ -0,0 +1,172 @@
+            Flags: d3d12::D3D12_DESCRIPTOR_HEAP_FLAG_NONE,
+            NodeMask: 0,
+        };
+
+        let mut heap: *mut d3d12::ID3D12DescriptorHeap = ptr::null_mut();
+        let handle_size = unsafe {
+            device.CreateDescriptorHeap(
+                &desc,
+                &d3d12::ID3D12DescriptorHeap::uuidof(),
+                &mut heap as *mut *mut _ as *mut *mut _,
+            );
+            device.GetDescriptorHandleIncrementSize(ty) as usize
+        };
+
+        let start = unsafe { (*heap).GetCPUDescriptorHandleForHeapStart() };
+
+        HeapLinear {
+            handle_size,
+            num: 0,
+            size,
+            start,
+            _raw: unsafe { ComPtr::from_raw(heap) },
+        }
+    }
+
+    pub fn alloc_handle(&mut self) -> d3d12::D3D12_CPU_DESCRIPTOR_HANDLE {
+        assert!(!self.full());
+
+        let slot = self.num;
+        self.num += 1;
+
+        d3d12::D3D12_CPU_DESCRIPTOR_HANDLE {
+            ptr: self.start.ptr + self.handle_size * slot,
+        }
+    }
+
+    pub fn full(&self) -> bool {
+        self.num >= self.size
+    }
+
+    pub fn clear(&mut self) {
+        self.num = 0;
+    }
+}
+
+const HEAP_SIZE_FIXED: u32 = 64;
+
+// Fixed-size free-list allocator for CPU descriptors.
+struct Heap {
+    // Bit flag representation of occupied handles in the heap.
+    //
+    //  0 - Occupied
+    //  1 - free
+    occupancy: u64,
+    handle_size: usize,
+    start: d3d12::D3D12_CPU_DESCRIPTOR_HANDLE,
+    _raw: ComPtr<d3d12::ID3D12DescriptorHeap>,
+}
+
+impl Heap {
+    pub fn new(device: &ComPtr<d3d12::ID3D12Device>, ty: d3d12::D3D12_DESCRIPTOR_HEAP_TYPE) -> Self {
+        let desc = d3d12::D3D12_DESCRIPTOR_HEAP_DESC {
+            Type: ty,
+            NumDescriptors: HEAP_SIZE_FIXED,
+            Flags: d3d12::D3D12_DESCRIPTOR_HEAP_FLAG_NONE,
+            NodeMask: 0,
+        };
+
+        let mut heap: *mut d3d12::ID3D12DescriptorHeap = ptr::null_mut();
+        let handle_size = unsafe {
+            device.CreateDescriptorHeap(
+                &desc,
+                &d3d12::ID3D12DescriptorHeap::uuidof(),
+                &mut heap as *mut *mut _ as *mut *mut _,
+            );
+            device.GetDescriptorHandleIncrementSize(ty) as usize
+        };
+        let start = unsafe { (*heap).GetCPUDescriptorHandleForHeapStart() };
+
+        Heap {
+            handle_size,
+            occupancy: !0, // all free!
+            start,
+            _raw: unsafe { ComPtr::from_raw(heap) },
+        }
+    }
+
+    pub fn alloc_handle(&mut self) -> d3d12::D3D12_CPU_DESCRIPTOR_HANDLE {
+        assert!(!self.full());
+
+        // Find first free slot.
+        let slot = self.occupancy.trailing_zeros() as usize;
+        // Set the slot as occupied.
+        self.occupancy ^= 1 << slot;
+
+        d3d12::D3D12_CPU_DESCRIPTOR_HANDLE {
+            ptr: self.start.ptr + self.handle_size * slot,
+        }
+    }
+
+    pub fn full(&self) -> bool {
```

Comment:
given that this is not an accessor, I don't think it's idiomatic to omit the verb here. Should be `is_full()` to mimic `is_empty()` in standard collections